apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java

```diff
@@ -251,7 +270,50 @@ private String readValue() throws IOException {
                     ch = readChar();
                 }
                 return s;
-            } else if (ch == '\n' || ch < 0 || ch == '\r') {
+            } else if (ch == textFieldCharacter) {
+                int state=1;
+                boolean endLex=false;
+                while (!endLex) {
+                    ch = readChar();
+                    switch (state) {
+                        case 0: //Error state
+                            break;
+                        case 1:
+                            if (ch == textFieldCharacter) {
+                                state = 2;
+                            } else if (isEOL(ch)) {
+                                state = 0;//error state
+                                endLex = true;
+                                endOfLine = true;
+                            }
+                            break;
+                        case 2:
+                            if (ch == textFieldCharacter) {
+                                state = 1;
+                            } else if (ch == fieldSeparatorRead) {
+                                state=3;//end state
+                                endLex = true;
+                            } else if (endOfFile=isEOL(ch)) {
+                                state=3;//end state
+                                endLex = true;
+                                endOfLine = true;
+                            } else if (isWhiteSpace(ch)) {
+                                //ignore
+                            } else {
+                                state = 0;
+                                endLex = true;
+                            }
+                            break;
+                    }
+                }
+                String s = new String(inputBuffer,
+                        inputBufferStart, inputBufferPos - inputBufferStart - 1);
+                if (!preserveWhitespace) {
+                    s = s.trim();
+                }
+                inputBufferStart = -1;
+                return s;
```

Comment:
Just to think about
Is it possible to extract it to a separate method with returning result? 
My concern is that this method is going to be too complex.